subsys/net/lib/lwm2m/lwm2m_pull_context.c

```diff
@@ -344,6 +359,18 @@ static void firmware_transfer(struct k_work *work)
 	int ret;
 	char *server_addr;
 
+	struct k_delayed_work *delayed_work;
+
+	delayed_work = CONTAINER_OF(work, struct k_delayed_work, work);
+
+	ret = k_sem_take(&lwm2m_pull_sem, K_NO_WAIT);
```

Comment:
Isn't it a bit of a waste, to keep separate PULL context per each object instance, but then limit the usage to a single download at once (note that this structure is pretty large due to fields like `uri` or `lwm2m_ctx`)?  Perhaps it'd be more economic to have a single context, and then verify inside `lwm2m_pull_context_start_transfer()` if it's in use (it'd then be the caller's responsibility to try again later).

Reply:
Agree. I'll go single context.

Reply:
@rlubos I would appreciate you taking a look at this solution https://github.com/zephyrproject-rtos/zephyr/pull/40552/commits/afbb37219139acd2467602f543c850be9e6ed23e - this is handling failure when a package is already being pulled.

Safety check if the work is already scheduled might be unnecessary but I'd really like your input on this.

Reply:
I just wonder, do we really need `firmware_work` in this case? This initial check to determine if download is already ongoing looks fine, but I don't see why would need a work here, `firmware_transfer()` could just be a regular function called from that context.

I'd rather see a delayed work use on the caller side - if the function returns `-EALREADY` it should set a delayed work and retry later.

I also don't think we should expect `struct firmware_pull_context *` as an argument for the `lwm2m_pull_context_start_transfer()` - since the only instance in use would exist in the `lwm2m_pull_context.c`, the callers don't even need to be aware of this structure. We only need to provide a few information needed to start the download - URI string pointer, instance ID, callback pointers, perhaps nicely wrapped into some helper struct. But this struct does not need to be in the static storage, it could exist on a stack, since all of the data would be copied into the pull context when the download is started.

Reply:
Hard to disagree with anything you've pointed out here.

I'll get to implementing these changes.